### PR TITLE
Wavelength calibration in multi-slit mode

### PR DESCRIPTION
**Pull request by rcooke-ast**
This new instalment of the wavelength calibration algorithm features a stretch+shift and cross correlation of all slits to improve the wavelength calibration (particularly of those slits where the wavelength calibration fails in the pattern matching). This new code is primarily used for multi-slit data.

This branch passes all of the tests in the dev suite (just waiting to confirm the DEIMOS multi's pass).

### CODE REVIEW

File: pypeit/core/wavecal/autoid.py [hunk 1]
```diff
@@ -502,42 +506,224 @@ def run(self):
...
                           '  Final RMS of fit             = {:g}'.format(best_final_fit['rms']) + msgs.newline() +
                           '---------------------------------------------------')
+                self._all_patt_dict[str(slit)] = copy.deepcopy(best_patt_dict)
+                self._all_final_fit[str(slit)] = copy.deepcopy(best_final_fit)
+
+        # Now that all slits have been inspected, cross match to generate a
+        # master list of all lines in every slit, and refit all spectra
+        if self._nslit > 1:
+            obad_slits = self.cross_match(good_fit)
+            while obad_slits.size > 0:
```
Comment (profxj) [inline pypeit/core/wavecal/autoid.py]: Is it possible to get into an infinite loop here?
Just checking.

File: pypeit/core/wavecal/autoid.py [hunk 2]
```diff
@@ -502,42 +506,224 @@ def run(self):
...
+                            # This is probably not a good match
+                            continue
+                        lindex = np.append(lindex, np.argmin(bstwv))
+                        dindex = np.append(dindex, dd)
+            # Finalize the best guess of each line
+            # Initialise the patterns dictionary
+            patt_dict = dict(acceptable=False, nmatch=0, ibest=-1, bwv=0., min_ampl=self._min_ampl,
+                             mask=np.zeros(bsdet.size, dtype=np.bool))
+            patt_dict['sign'] = sign
+            patt_dict['bwv'] = np.median(wcen[np.where(wcen != 0.0)])
```
Comment (profxj) [inline pypeit/core/wavecal/autoid.py]: could probably right the RHS as:

```np.median(wcen[wcen != 0.])```

Comment (rcooke-ast) [reply pypeit/core/wavecal/autoid.py]: Done

**Comment by rcooke-ast** [comment]
I've added two hooks for the while loops to break (after 100 iterations, which is probably excessive).

**Comment by rcooke-ast** [comment]
The dev suite passes (to the same extent as before)... merging.

**Pull request merged**